### PR TITLE
TST: start testing nightlies on CPython 3.12

### PR DESCRIPTION
**Pull request by neutrinoceros**
matplotlib started shipping (unstable) wheels on pypi.anaconda.org, let's try and see where we're at.

edit: no they didn't, I just misread [the index](https://pypi.anaconda.org/scientific-python-nightly-wheels/simple/matplotlib/). I'll refresh this PR when it actually happens.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
to follow upstream progress, see https://github.com/matplotlib/matplotlib/pull/24711

**Comment by neutrinoceros** [comment]
Test jobs appear to be "cancelled" mid-air, which is worrying; I'd much rather have to deal with actual failures. I don't have a clue what's happening yet, but I'm not excluding the possibility that it's just a temporary fluke on GitHub's runners, so maybe I'll try again later.

**Comment by neutrinoceros** [comment]
I've been trying to skip tests that I found hanged or crashed pytest but it doesn't seem practical. There are a million moving pieces and as many possible reasons this wouldn't work at the moment, so I'll just try again in a couple weeks when hopefully more upstream dependencies have published some wheels.

**Comment by neutrinoceros** [comment]
this is currently based onto #4652

File: yt/geometry/coordinates/tests/test_cartesian_coordinates.py [hunk 1]
```diff
@@ -13,19 +13,39 @@ def test_cartesian_coordinates():
...
     axes = sorted(set(ds.coordinates.axis_name.values()))
-    for i, axis in enumerate(axes):
-        dd = ds.all_data()
+    print("axes set", flush=True)
+
+    dd = ds.all_data()
+    print("dd loaded", flush=True)
+
+    i = 0
+    for axis in axes:
```
Comment (matthewturk) [inline yt/geometry/coordinates/tests/test_cartesian_coordinates.py:24]: I don't like this change.  I'm surprised it is necessary for Python 3.12.

Comment (neutrinoceros) [reply yt/geometry/coordinates/tests/test_cartesian_coordinates.py:24]: Like most changes currently bundled in this draft PR, it's not. This is a temporary commit, and a desperate attempt to debug this test. At the moment, I have *no* idea why it crashes, and I haven't been able to reproduce the problem locally.

Comment (matthewturk) [reply yt/geometry/coordinates/tests/test_cartesian_coordinates.py:24]: :+1:

**Comment by neutrinoceros** [comment]
currently based off of #4641

**Comment by neutrinoceros** [comment]
As we're still having issues with Python 3.12.0rc3, I reported the crash upstream https://github.com/python/cpython/issues/109602

**Comment by neutrinoceros** [comment]
bleeding-edge tests are finally green thanks to @yut23's patch in Cython !

File: .github/workflows/bleeding-edge.yaml [hunk 2]
```diff
@@ -37,23 +37,30 @@ jobs:
     - name: Set up Python (newest testable version)
       uses: actions/setup-python@v4
       with:
-        # the '-dev' suffix allows to use alphas and betas if no final release is available yet
-        # this version should be upgraded as often as possible, typically once a year when numpy
-        # and Cython are known to be compatible
-        python-version: '3.11-dev'
+        # the '-dev' suffix allows to use pre-releases if no final release is available yet
+        # this version should be upgraded as often as possible, typically once a year when
+        # and Cython, numpy and matplotlib are known to be compatible
```
Comment (cphyc) [inline .github/workflows/bleeding-edge.yaml:42]:
```suggestion
        # this version should be upgraded as often as possible, typically once a year when
        # Cython, numpy and matplotlib are known to be compatible
```

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Comment by neutrinoceros** [comment]
@yt-fido test this please

**Review by Xarthisius — APPROVED** [review]

**Pull request merged**